data/basics/testing/roundtrip/roundtrip.go

```diff
@@ -0,0 +1,197 @@
+	return multiOption{randomCountOption(count), randomOptsOption(opts)}
+}
+
+// NoRandomCases disables RandomizeObject testing (but still runs NearZeros).
+// Use this when RandomizeObject generates invalid values for constrained types.
+// Combine with NoNearZeros() to disable all automatic testing.
+func NoRandomCases() CheckOption {
+	return randomCountOption(0)
+}
+
+type skipNearZerosOption struct{}
+
+func (skipNearZerosOption) apply(cfg *checkConfig) {
+	cfg.skipNearZeros = true
+}
+
+// NoNearZeros disables NearZeros testing, only using RandomizeObject for random variants.
+// Use this for non-struct types (maps, slices) where NearZeros doesn't apply.
+func NoNearZeros() CheckOption {
+	return skipNearZerosOption{}
+}
+
+// WithRapid specifies a rapid.Generator to use for property-based testing.
+// If provided, rapid.Check will be used instead of protocol.RandomizeObject (runs 100 tests).
+func WithRapid[A any](gen *rapid.Generator[A]) CheckOption {
+	return rapidGenOption{gen: gen}
+}
+
+type multiOption []CheckOption
+
+func (m multiOption) apply(cfg *checkConfig) {
+	for _, opt := range m {
+		opt.apply(cfg)
+	}
+}
+
+// Check verifies that converting from A -> B -> A yields the original value.
+// By default, tests the provided example, all NearZeros variants (one per field),
+// and 100 randomly generated values using protocol.RandomizeObject.
+// Use WithRapid to provide a custom rapid.Generator for property-based testing.
+// Use NoRandomCases to disable RandomizeObject (still runs NearZeros).
+// Use NoNearZeros to disable NearZeros (for non-struct types like maps).
+// Use Opts to customize the number of random tests or pass RandomizeObjectOptions.
+func Check[A any, B any](t *testing.T, a A, toB func(A) B, toA func(B) A, opts ...CheckOption) bool {
+	cfg := checkConfig{}
+	for _, opt := range opts {
+		opt.apply(&cfg)
+	}
+
+	// Test the provided example first
+	if !checkOne(t, a, toB, toA) {
+		t.Errorf("Round-trip failed for provided example: %+v", a)
+		return false
+	}
+
+	// Use rapid property testing if generator provided
+	if cfg.useRapid {
+		gen, ok := cfg.rapidGen.(*rapid.Generator[A])
+		if !ok {
+			t.Errorf("Invalid rapid generator type")
+			return false
+		}
+
+		// Run rapid property tests (runs 100 tests by default)
+		// Note: rapid.Check controls the count, not us
+		passed := true
+		rapid.Check(t, func(t1 *rapid.T) {
+			randA := gen.Draw(t1, "value")
+			if !checkOne(t, randA, toB, toA) {
+				t.Errorf("Round-trip failed for rapid-generated value: %+v", randA)
+				passed = false
+			}
+		})
+		return passed
+	}
+
+	// Test NearZeros (one test per field) - comprehensive and deterministic
+	// Skip if explicitly disabled
+	if !cfg.skipNearZeros {
+		nearZeroValues := NearZeros(t, a)
+		for i, nzA := range nearZeroValues {
+			if !checkOne(t, nzA, toB, toA) {
+				t.Errorf("Round-trip failed for NearZero variant %d: %+v", i, nzA)
+				return false
+			}
+		}
+	}
```

Comment:
I'd probably put `NearZeros` cases first, because when they fail and print, the field that's the problem is very, very obvious.